internal/errorcodes/errorcodes.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implieout.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errorcodes
+
+import (
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+type Reason string
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+)
+
+var StateETagMismatchReason = Reason("DAPR_STATE_ETAG_MISMATCH")
+
+type ResourceInfoData struct {
+	ResourceType, ResourceName string
+}
+
+func FeatureEnabled(md map[string]string) bool {
+	if _, ok := md[ErrorCodesFeatureMetadataKey]; ok {
+		return true
+	}
+	return false
+}
+
+// NewStatusError returns a Status representing Code, error Reason, Message, and optional ResourceInfo and Metadata.
+// When successful, it returns a StatusError, otherwise returns the original error
+func NewStatusError(code codes.Code, err error, errDescription string, reason Reason, rid *ResourceInfoData, metadata map[string]string) error {
```

Comment:
Why do we need both `err error` and `errDescription string`? Can't the two be inferred from one another?

Reply:
The idea behind passing both separate is to give the calling code flexibility on the `errDescription`.

Reply:
I like your suggestion better. will change it.

Reply:
I see that in the code you've written so far, in most cases the error is similar or derived.

how about creating a custom error object that has a "Description()" property that is optional?

```go
type ComponentError struct {
    err error
	code codes.Code
	description string
	resourceInfoData *ResourceInfoData
	metadata map[string]string
}

func NewComponentError(code codes.Code, err error) ComponentError {
	return ComponentError{
		code: code,
		err: err,
	}
}

// Error implements the error interface.
func (c ComponentError) Error() string {
	return c.err.Error()
}

// Unwrap implements the error unwrapping interface.
func (c ComponentError) Unwrap() error {
	return c.err
}

// Description returns the description of the error.
func (c ComponentError) Description() string {
	if c.description {
		return c.description
	}
	return c.err.Error()
}

func (c *ComponentError) SetDescription(description string) {
	c.description = description
}

// Add methods to access code, resourceInfoData, metadata
// Add methods to set resourceInfoData and metadata
```

Then you can pass a ComponetnError around as an `error`. In the code that handles errors, you can check:

```go
var compErr ComponentError
if errors.As(err, &compErr) {
// ...
}
```

Using custom errors is a more idiomatic thing in Go. It also allows accessing the `err` property for unwrapping.

Some links: https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go

Reply:
Yep, custom errors are idiomatic in Go, but the idea is to follow [The proposal and use gRPC error handling: 0009-BCIRS-error-handling-codes ](https://github.com/dapr/proposals/blob/main/0009-BCIRS-error-handling-codes.md)

Reply:
This is still compliant with the proposal :)

You could also add the GRPCStatus method to make the custom error be a gRPC error too: https://github.com/dapr/dapr/blob/master/pkg/messages/api_error.go#L79-L83

Reply:
I like the idea of having an internal struct to hold all the error information. On the other hand, I am not a fan of having it as a component specific thing. I suggest we have an internal error struct defined in kit and that being the common object for all dapr internal code. The runtime then will have a logic to transform that object into the gRPC error we agreed in the proposal.

Reply:
+1 on getting this in Kit. I like that